contracts/escrows/CountdownGriefingEscrow.sol

```diff
@@ -0,0 +1,357 @@
+pragma solidity ^0.5.0;
+
+import "../helpers/openzeppelin-solidity/math/SafeMath.sol";
+import "../helpers/openzeppelin-solidity/token/ERC20/IERC20.sol";
+import "../agreements/CountdownGriefing.sol";
+import "../modules/iFactory.sol";
+import "../modules/iRegistry.sol";
+import "../modules/Countdown.sol";
+import "../modules/Staking.sol";
+import "../modules/EventMetadata.sol";
+import "../modules/Operated.sol";
+import "../modules/Template.sol";
+
+/// @title CountdownGriefingEscrow
+/// @author Stephane Gosselin (@thegostep) for Numerai Inc
+/// @dev Security contact: security@numer.ai
+contract CountdownGriefingEscrow is Countdown, Staking, EventMetadata, Operated, Template {
+
+    using SafeMath for uint256;
+
+    Data private _data;
+    struct Data {
+        address buyer;
+        address seller;
+        address operator;
+        uint128 paymentAmount;
+        uint128 stakeAmount;
+        EscrowStatus status;
+        AgreementParams agreementParams;
+    }
+
+    struct AgreementParams {
+        uint120 ratio;
+        uint8 ratioType;
+        uint128 countdownLength;
+    }
+
+    event Initialized(
+        address buyer,
+        address seller,
+        address operator,
+        uint256 paymentAmount,
+        uint256 stakeAmount,
+        uint256 countdownLength,
+        bytes metadata,
+        bytes agreementParams
+    );
+    event DataSubmitted(bytes data);
+    event StakeDeposited(address seller, uint256 stake);
+    event PaymentDeposited(address buyer, uint256 payment);
+    event Finalized();
+    event Cancelled();
+
+    /// @notice Constructor
+    /// @dev Access Control: only factory
+    ///      State Machine: before all
+    function initialize(
+        address buyer,
+        address seller,
+        address operator,
+        uint256 paymentAmount,
+        uint256 stakeAmount,
+        uint256 escrowCountdown,
+        bytes memory metadata,
+        bytes memory agreementParams
+    ) public initializeTemplate() {
+        // set participants if defined
+        if (buyer != address(0))
+            _data.buyer = buyer;
+        if (seller != address(0))
+            _data.seller = seller;
```

Comment:
Currently, buyer and seller can't be the same address, since we check Staking.getStake(_data.seller/buyer) in many places.

Reply:
I think the only way around this issue will be to have a boolean flag that gets flipped when the stake / payment is deposited

Reply:
fixed in 50df9d5